Add foundations for extensible validation in forms (#6161)
plone/volto: docs/source/configuration/validation.md

---
myst:
  html_meta:
    "description": "Client side form field validation"
    "property=og:description": "Client side form field validation"
    "property=og:title": "Form fields validation"
    "keywords": "Volto, Plone, frontend, React, configuration, form, fields, validation"
---

# Client side form field validation

Volto provides a mechanism for delivering form field validation in an extensible way.
This extensibility is based on the Volto component registry.

## Registering a validator

You can register a validator using the component registry API from your add-on configuration.
All validators are registered under the name `fieldValidator`.
The validators are registered using the `dependencies` array of the `registerComponent` API to differentiate the kind of validator to be registered.

### `default` validators

These validators are registered and applied to all fields.

```ts
config.registerComponent({
  name: 'fieldValidator',
  dependencies: ['default', 'minLength'],
  component: minLengthValidator,
});
```

It takes two `dependencies`.

> The first element should be the `default` identifier, and the second you can set it up to identify the validator.

Reviewer: It's not explained why `default` should be the first identifier.

In the case of the example, this other dependency is `minLength`.
It can be any string.

### Per field `type` validators

These validators are applied depending on the specified `type` of the field.

```ts
config.registerComponent({
  name: 'fieldValidator',

>   dependencies: ['integer', 'maximum'],

Reviewer: Here the `default` is missing.

Author: ummm, in that line, the `default` is not needed here, because it's not a default validator, it's a `type` one, so the first dependency is the name of the type.

This type of validator only applies to content-type validators.

### Per block type and field name validator

These validators are applied depending on the block type in combination with the name of the field in the block settings JSON schema.

```ts
config.registerComponent({
  name: 'fieldValidator',
  dependencies: ['slider', 'url'],
  component: urlValidator,
});
```
The first dependency should be the `id` of the block, and the second the `id` of the field.
This type of validator only applies to blocks.

### Specific validator using the `validator` key in the field

A final type of validator is applied to the field if the `validator` key is present in the JSON schema definition of the form field.

```ts
config.registerComponent({
  name: 'fieldValidator',
  dependencies: ['isURL'],
  component: urlValidator,
});
```

The dependencies take one single name, in this case, the name of the validator.
You should specify the validator in the JSON schema of the block (or as additional data in the content type definition).

```ts
let blockSchema = {
  properties: {
    ...schema.properties,
    customField: {
      title: 'Default field',
      description: '',

>       validator: 'isURL',

Reviewer: Maybe allow passing directly a validator function as well? We're way past the point "schemas should be serializable" and for one-offs it's easier to declare them inline.

Author: If the pattern we are teaching is go for "named" components and call them by name (partially because of the duality forms from the backend/content types and the client defined forms in blocks), I'd refrain to do so, because then we will have two use cases, one where you can do it, and another use case in which you can't. 

Also, in terms of explaining this in docs will became tricky and separate this is already confusing (I've been there when reviewing this with @ericof). Some other considerations is testability, because in that regard, modularization >> inlining things.

So, in principle, I would be -1 for this.

---
myst:
  html_meta:
    "description": "Client side form field validation"
    "property=og:description": "Client side form field validation"
    "property=og:title": "Form fields validation"
    "keywords": "Volto, Plone, frontend, React, configuration, form, fields, validation"
---

# Client side form field validation

Volto provides a mechanism for delivering form field validation in an extensible way.
This extensibility is based on the Volto component registry.

> ## Registering a validator

Reviewer 2: ```suggestion
## Register a validator
```

You can register a validator using the component registry API from your add-on configuration.
All validators are registered under the name `fieldValidator`.
The validators are registered using the `dependencies` array of the `registerComponent` API to differentiate the kind of validator to be registered.

### `default` validators

These validators are registered and applied to all fields.

```ts
config.registerComponent({
  name: 'fieldValidator',
  dependencies: ['default', 'minLength'],
  component: minLengthValidator,
});
```

It takes two `dependencies` being the first element a fixed `default` identifier, and the second you can set it up to identify the validator itself.
In the case of the example, this other dependency is `minLength`.
It can be any string.

> ### Per field `type` validators

Reviewer 2: ```suggestion
### Field `type` validators
```

> These validators are applied depending on the specified `type` of the field.

Reviewer 2: ```suggestion
Field `type` validators are applied according to the specified `type` of the field.
```

```ts
config.registerComponent({
  name: 'fieldValidator',
  dependencies: ['integer', 'maximum'],
  component: maximumValidator,
});
```

It takes two `dependencies` since we can potentially have several validators for the same `type`.
The first element should be the field `type`, and the second you should set it up to identify the validator itself.
You should specify the `type` in the JSON schema of the block (in a content type, it is included in the default serialization of the field).
If a field does not specify type, it assumes a `string` type as validator.

> The next example is for the use case of JSON schema defined in a block:

Reviewer 2: ```suggestion
The following example shows how to use a JSON schema defined in a block.
```

```ts
let blockSchema = {
  // ... fieldset definition in here
  properties: {
    ...schema.properties,
    customField: {
      title: 'My custom field',
      description: '',
      type: 'integer',
    },
  },
  required: [],
};
```

> ### Per field `widget` validators

Reviewer 2: ```suggestion
### Field `widget` validators
```

> These validators are applied depending on the specified `widget` of the field.

Reviewer 2: ```suggestion
Field `widget` validators are applied according to the specified `widget` for the field.
```

```ts
config.registerComponent({
  name: 'fieldValidator',
  dependencies: ['phoneNumber', 'isValidPhone'],
  component: phoneValidator,

> });

Reviewer 3: What makes this a widget validator instead of a type validator?

```

It takes two `dependencies` since we can potentially have several validators for the same `widget`.
The first element should be the name of the `widget`, and the second you can set it up to identify the validator.
You should specify the `widget` in the JSON schema of the block (or as additional data in the content type definition).
The next example is for the use case of a block JSON schema:

```ts
let blockSchema = {
  // ... fieldset definition in here
  properties: {
    ...schema.properties,
    phone: {
      title: 'Phone number',
      description: '',
      widget: 'phoneNumber',
    },
  },
  required: [],
};
```

> ### Per behavior and field name validator

Reviewer 2: ```suggestion
### Behavior and field name validator
```

> These validators are applied depending on the behavior (usually coming from a content type definition) in combination with the name of the field.

Reviewer 2: ```suggestion
Behavior and field name validators are applied according to the behavior combined with the name of the field.
The behavior usually comes from a content type definition.
```

```ts
config.registerComponent({
  name: 'fieldValidator',
  dependencies: ['plone.eventbasic', 'start', 'startValidator'],
  component: startEventDateRangeValidator,
});
```

> The first dependency should be the name of the behavior, and the second the name (`id`) of the field.

Reviewer 2: ```suggestion
The `dependencies` key accepts an array of two or three values.
The first element is a string that is the name of the behavior.
The second element is the name (`id`) of the field.
```

> It can get a third dependency in case you want to specify several validators for the same behavior - field id combination.

Reviewer 2: ```suggestion
If you want to specify several validators for the same behavior, you can provide an optional third dependency as a field id combination.
```

> This type of validator only applies to content-type validators.

Reviewer 2: ```suggestion
This type of validator applies only to content type validators.
```

> ### Per block type and field name validator

Reviewer 2: ```suggestion
### Block type and field name validator
```

> These validators are applied depending on the block type in combination with the name of the field in the block settings JSON schema.

Reviewer 2: ```suggestion
Block type and field name validators are applied according to the block type combined with the name of the field in the block settings JSON schema.
```

```ts
config.registerComponent({
  name: 'fieldValidator',
  dependencies: ['slider', 'url', 'isURL'],
  component: urlValidator,
});

> ```

Reviewer 2: Move example after description.

> The first dependency should be the `id` of the block, and the second the `id` of the field.

Reviewer 3: Isn't it the `@type` of the block?

> The first dependency should be the `id` of the block, and the second the `id` of the field.

Reviewer 2: ```suggestion
The `dependencies` key accepts an array of two values.
The first element is a string that is the `type` of the block.
The second element is the `id` of the field.
```

> It can get a third dependency in case you want to specify several validators for the same block type - field id combination.

Reviewer 2: ```suggestion
To specify several validators for the same block type, you can provide an optional third value as a field id combination.
```

> This type of validator only applies to blocks.

Reviewer 2: ```suggestion
This type of validator applies only to blocks.
```

### Specific validator using the `validator` key in the field

> A final type of validator is applied to the field if the `validator` key is present in the JSON schema definition of the form field.

Reviewer 2: ```suggestion
You can apply a specific validator to a field by using the `validator` key in the JSON schema definition of the form field.
```

```ts
config.registerComponent({
  name: 'fieldValidator',
  dependencies: ['isURL'],
  component: urlValidator,
});

> ```

Reviewer 2: Move after description.

> The dependencies take one single name, in this case, the name of the validator.

Reviewer 2: ```suggestion
The `dependencies` key accepts an array with a single value, the name of the validator.
```

> You should specify the validator in the JSON schema of the block (or as additional data in the content type definition).

Reviewer 2: ```suggestion
You should specify the validator either in the JSON schema of the block or as additional data in the content type definition.
```

```ts
let blockSchema = {
  // ... fieldset definition in here
  properties: {
    ...schema.properties,
    customField: {
      title: 'Default field',
      description: '',
      validator: 'isURL',
    },
  },
  required: [],
};

> ```

Reviewer 2: Move example after description.

It does not need to be tied to any field `type` or `widget` definition.

> It runs in addition to all the above, so it complements the other validators if any apply.

Reviewer 3: My biggest piece of feedback on this overall is using an array for the `dependencies` feels like trying to fit a square peg into a round hole. It's quite hard to keep track of all the different ways validators can be registered and which values need to go in the dependencies for each one. And it also seems to be like there might be some ambiguous cases, like if there is a field type and a widget type with the same name.

I don't know if the registry supports this but would it be possible to make things a bit more explicit by specifying the dependencies as an object instead of an array? So for example:
```
config.registerComponent({
  name: 'fieldValidator',
  dependencies: {
    blockType: 'slider',
    fieldName: 'url',
    validatorName: 'isURL',
  },
  component: urlValidator,
});
```

Author: @davisagli TBH, I'm not fully satisfied with it... I thought that the reason was mainly because we can't get it to work as satisfying as the ZCA works :) I envisioned the resolution algorithm using a string and concating the deps if we have an array, so I've played with that concept, even in the new `getComponents` registry API method.

Using an object would be possible too, I guess. I can play with the idea, and leave the original API in place also.

Author: @davisagli I did an initial implementation of using an object instead of an array but I have doubts on how to hash the resultant key to store that object, so then `getComponents` work in the same simple way, under the same concepts that we had before. It does work, but under some assumptions that I do not like or they are not satisfactory/safe :/ 

Maybe if you have some time, we can talk about it.

## Volto's default validators

> Volto provides a set of validators by default, you can find them in this module: `packages/volto/src/config/validators.ts`

Reviewer 2: ````suggestion
Volto provides a set of validators.
You can find them in the module {file}`packages/volto/src/config/validators.ts`.

```{todo}
Document the API as a reference in another file, after the dust settles.
```
````

> ### How to override them

Reviewer 2: ```suggestion
### Override Volto's default validators
```

> You can override them in your add-on as any other component defined in the registry, by redefining them using the same `dependencies`, and providing your own.

Reviewer 2: ````suggestion
You can override Volto's default validators in your add-on as any other component defined in the registry.
To do so, use the `dependencies` key, and provide your own values.

```{todo}
Provide usage example.
```
````

## Signature of a validator

A validator has the following signature:

```ts
type Validator = {
  // The field value
  value: string;
  // The field schema definition object
  field: Record<string, any>;
  // The form data
  formData?: any;
  // The intl formatMessage function
  formatMessage: Function;
};
```

> This is an example of an `isNumber` validator:

Reviewer 2: ```suggestion
The following example shows an `isNumber` validator.
```

```ts
export const isNumber = ({ value, formatMessage }: Validator) => {
  const floatRegex = /^[+-]?\d+(\.\d+)?$/;
  const isValid =
    typeof value === 'number' && !isNaN(value) && floatRegex.test(value);
  return !isValid ? formatMessage(messages.isNumber) : null;
};
```

> Using the `formData` you can perform validation checks using other field data as source.

Reviewer 2: ```suggestion
Using `formData`, you can perform validation checks between multiple fields' data.
```

> This is interesting in the case that two fields are related, like `start` and `end` dates.

Reviewer 2: ````suggestion
This is useful when you need to compare the values of multiple fields, such as ensuring that the end date and time of an event is after its start.

```{todo}
Provide usage example.
```
````
